Temporal index select

taobrienlbl opened the pull request:
Wrote `teca_temporal_index_select`: an algorithm for selecting specific time indices.  Added as an optional stage in `teca_temporal_reduction`.

taobrienlbl commented:
@burlen - the only remaining thing to change here is the python interface for `set_indices()`, which we have been e-mailing about.  Once I figure that out, I'll push it up and squash; otherwise, I think this is ready to review.

burlen commented on alg/teca_temporal_index_select.cxx:
```diff
@@ -0,0 +1,292 @@
+        }
+
+        // set the reported time extent to be the same as the request
+        std::vector<size_t> index_extent;
```
would be more efficient to use `size_t index_extent[2]`

burlen commented on alg/teca_temporal_index_select.cxx:
```diff
@@ -0,0 +1,292 @@
+        // set the reported time extent to be the same as the request
+        std::vector<size_t> index_extent;
+        request.get(key, index_extent);
+        md_out.set(key, index_extent);
```
There are also some convenience methods to do this.
https://teca.readthedocs.io/en/latest/doxygen/classteca__dataset.html#af9276ab5dc2067a00f767f811a7c8c7d

taobrienlbl replied:
to make sure I understand, you're suggesting replacing lines 284-287 with `out_mesh->set_request_indices(key, index_extent)`?

burlen replied:
Yes.

taobrienlbl replied:
cool, thanks for clarifying!

burlen reviewed (commented):
Nice! I saw a few things that we might want to address.

taobrienlbl commented on alg/teca_temporal_index_select.cxx:
```diff
@@ -0,0 +1,292 @@
+
+                const NT1 *p_t_in = sp_t_in.get();
+                // set this time value
+                t_out->set(i, p_t_in[this->indices[i]]);
```
could you please expand on why it would be better not to use the virtual API here and in what cases the virtual API should be used?  I think there's a point about C++ code design that I'm unfamiliar with

burlen replied:
it's inefficient. the virtual api could be useful when you don't know the type.

The virtual methods on teca_variant_array handles all the casting internally, then routes to the correct device CPU or GPU. so these are the overheads here:  each call to virtual set has the cost of a function call  because virtual calls can't be in lined, then our code has to figure out the actual type, then the allocator, and finally set the value (this may result in another function call eg. cudaMemcpy). you're doing it in a loop so it happens multiple times.

Here you know the type. So instead you can cast to the typed array and use the `data()` method to get the pointer, then each setting each value is a single machine instruction. All of the allocator based routing is skipped. The case can be done outside the loop.

burlen replied:
as a rule of thumb: use the virtual methods when you can transfer all the data with a single call.

burlen replied:
one of the sad things about the changes to support GPUs is that calls to set or get a single value that were essentially free before, are now very very costly. Now we need to use bulk data transfers when ever possible.

taobrienlbl replied:
got it, thanks for the thorough explanation!

taobrienlbl commented:
Thanks for the review @burlen!  I addressed all the comments; thanks for the thorough examination.  @amandasd - if you have a moment to take a look before the weekend, I'd appreciate it!  If not, that's ok too: I can go ahead and merge with the feedback from Burlen.

amandasd commented:
Hi @taobrienlbl. I took a quick look at the code earlier today and it looks really good! Good job!

taobrienlbl commented:
Thanks @amandasd !  Would you mind hitting "approve"?
